File: src/viewmodels/courses/QuizViewModel.js
// src/viewmodels/courses/QuizViewModel.js
import { useState, useEffect } from "react";
import { auth } from "../../firebase";
import QuizModel from "../../models/QuizModel";

export const useQuizViewModel = (courseId, chapterId) => {
  const [quiz, setQuiz] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  
  // État du quiz en cours
  const [currentQuestionIndex, setCurrentQuestionIndex] = useState(0);
  const [selectedAnswers, setSelectedAnswers] = useState({});
  const [showResults, setShowResults] = useState(false);
  const [score, setScore] = useState(0);
  const [quizStarted, setQuizStarted] = useState(false);
  const [submitting, setSubmitting] = useState(false);

  const user = auth.currentUser;

  // Charger le quiz
  useEffect(() => {
    const loadQuiz = async () => {
      try {
        setLoading(true);
        const quizData = await QuizModel.getQuizForChapter(courseId, chapterId);
        
        if (!quizData || !quizData.questions || quizData.questions.length === 0) {
          setError("Aucun quiz disponible pour ce chapitre");
          return;
        }

        setQuiz(quizData);
      } catch (err) {
        console.error("Erreur chargement quiz:", err);
        setError(err.message || "Erreur lors du chargement du quiz");
      } finally {
        setLoading(false);
      }
    };

    if (courseId && chapterId) {
      loadQuiz();
    }
  }, [courseId, chapterId]);

  // Démarrer le quiz
  const startQuiz = () => {
    setQuizStarted(true);
    setCurrentQuestionIndex(0);
    setSelectedAnswers({});
    setShowResults(false);
    setScore(0);
  };

  // Sélectionner une réponse
  const selectAnswer = (questionId, answerIndex) => {
    setSelectedAnswers({
      ...selectedAnswers,
      [questionId]: answerIndex,
    });
  };

  // Question suivante
  const nextQuestion = () => {
    if (currentQuestionIndex < quiz.questions.length - 1) {
      setCurrentQuestionIndex(currentQuestionIndex + 1);
    }
  };

  // Question précédente
  const previousQuestion = () => {
    if (currentQuestionIndex > 0) {
      setCurrentQuestionIndex(currentQuestionIndex - 1);
    }
  };

  // Soumettre le quiz
  const submitQuiz = async () => {
    if (!user) {
      alert("Vous devez être connecté pour soumettre le quiz");
      return;
    }

    // Vérifier que toutes les questions ont une réponse
    const unansweredQuestions = quiz.questions.filter(
      (q) => selectedAnswers[q.id] === undefined
    );

    if (unansweredQuestions.length > 0) {
      alert(`Veuillez répondre à toutes les questions (${unansweredQuestions.length} restante(s))`);
      return;
    }

    setSubmitting(true);

    try {
      // Calculer le score
      let correctAnswers = 0;
      quiz.questions.forEach((question) => {
        if (selectedAnswers[question.id] === question.correctAnswer) {
          correctAnswers++;
        }
      });

      setScore(correctAnswers);

      const totalQuestions = quiz.questions.length;
      const percentage = Math.round((correctAnswers / totalQuestions) * 100);
      const passed = percentage >= (quiz.passingScore || 60);

      // Sauvegarder dans Firebase
      await QuizModel.saveQuizResult(
        user.uid,
        courseId,
        chapterId,
        correctAnswers,
        totalQuestions,
        passed
      );

      setShowResults(true);
    } catch (err) {
      console.error("Erreur soumission quiz:", err);
      alert("Erreur lors de la soumission du quiz. Veuillez réessayer.");
    } finally {
      setSubmitting(false);
    }
  };

  // Recommencer le quiz
  const retryQuiz = () => {
    startQuiz();
  };

  const currentQuestion = quiz?.questions?.[currentQuestionIndex];
  const totalQuestions = quiz?.questions?.length || 0;
  const percentage = totalQuestions > 0 ? Math.round((score / totalQuestions) * 100) : 0;
  const passed = percentage >= (quiz?.passingScore || 60);

  return {
    quiz,
    loading,
    error,
    quizStarted,
    currentQuestion,
    currentQuestionIndex,
    totalQuestions,
    selectedAnswers,
    showResults,
    score,
    percentage,
    passed,
    submitting,
    startQuiz,
    selectAnswer,
    nextQuestion,
    previousQuestion,
    submitQuiz,
    retryQuiz,
  };
};

File: lib/viewmodels/quiz_viewmodel.dart
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart';
import '../models/course_model.dart';
import '../services/quiz_service.dart';

enum QuizState { loading, error, intro, questions, results }

class QuizViewModel extends ChangeNotifier {
  final QuizService _quizService = QuizService();
  final String courseId;
  final String chapterId;

  QuizModel? _quiz;
  QuizModel? get quiz => _quiz;

  QuizState _state = QuizState.loading;
  QuizState get state => _state;

  String? _error;
  String? get error => _error;

  int _currentQuestionIndex = 0;
  int get currentQuestionIndex => _currentQuestionIndex;

  Map<String, int> _selectedAnswers = {};
  Map<String, int> get selectedAnswers => _selectedAnswers;

  int _score = 0;
  int get score => _score;
  
  bool _passed = false;
  bool get passed => _passed;

  bool _submitting = false;
  bool get submitting => _submitting;

  QuizViewModel({required this.courseId, required this.chapterId}) {
    _loadQuiz();
  }

  Future<void> _loadQuiz() async {
    try {
      _state = QuizState.loading;
      notifyListeners();

      final quizData = await _quizService.getQuizForChapter(courseId, chapterId);

      if (quizData == null || quizData.questions.isEmpty) {
        _error = "Aucun quiz disponible pour ce chapitre";
        _state = QuizState.error;
      } else {
        _quiz = quizData;
        _state = QuizState.intro;
      }
    } catch (e) {
      _error = e.toString();
      _state = QuizState.error;
    } finally {
      notifyListeners();
    }
  }

  void startQuiz() {
    _currentQuestionIndex = 0;
    _selectedAnswers = {};
    _score = 0;
    _state = QuizState.questions;
    notifyListeners();
  }

  void selectAnswer(String questionId, int answerIndex) {
    _selectedAnswers[questionId] = answerIndex;
    notifyListeners();
  }

  void nextQuestion() {
    if (_quiz != null && _currentQuestionIndex < _quiz!.questions.length - 1) {
      _currentQuestionIndex++;
      notifyListeners();
    }
  }

  void previousQuestion() {
    if (_currentQuestionIndex > 0) {
      _currentQuestionIndex--;
      notifyListeners();
    }
  }

  Future<void> submitQuiz(BuildContext context) async {
    final user = FirebaseAuth.instance.currentUser;
    if (user == null) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text("Vous devez être connecté pour soumettre le quiz")),
      );
      return;
    }

    // Vérifier réponses manquantes
    final unansweredCount = _quiz!.questions.where((q) => !_selectedAnswers.containsKey(q.id)).length;
    if (unansweredCount > 0) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text("Veuillez répondre à toutes les questions ($unansweredCount restantes)")),
      );
      return;
    }

    _submitting = true;
    notifyListeners();

    try {
      // Calcul du score
      int correctAnswers = 0;
      for (var question in _quiz!.questions) {
        if (_selectedAnswers[question.id] == question.correctAnswer) {
          correctAnswers++;
        }
      }

      _score = correctAnswers;
      final totalQuestions = _quiz!.questions.length;
      final percentage = (_score / totalQuestions) * 100;
      _passed = percentage >= _quiz!.passingScore;

      // Sauvegarde
      await _quizService.saveQuizResult(
        userId: user.uid,
        courseId: courseId,
        chapterId: chapterId,
        score: _score,
        totalQuestions: totalQuestions,
        passed: _passed,
      );

      _state = QuizState.results;
    } catch (e) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text("Erreur lors de la soumission: $e")),
      );
    } finally {
      _submitting = false;
      notifyListeners();
    }
  }

  void retryQuiz() {
    startQuiz();
  }

  QuestionModel? get currentQuestion => 
      _quiz != null && _quiz!.questions.isNotEmpty 
      ? _quiz!.questions[_currentQuestionIndex] 
      : null;
      
  int get totalQuestions => _quiz?.questions.length ?? 0;
  
  int get percentage => 
      totalQuestions > 0 ? ((_score / totalQuestions) * 100).round() : 0;
}
